src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2626,107 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
 }
 FCIMPLEND
 
+static BOOL HasOverriddenEqualsWorker(MethodTable* mt)
+{
+    MethodTable::IntroducedMethodIterator it(mt);
+    for (; it.IsValid(); it.Next())
+    {
+        if (strcmp(it.GetMethodDesc()->GetName(), "Equals") == 0)
+        {
+            return TRUE;
+        }
+    }    
+
+    return FALSE;
+}
+
+static BOOL HasOverriddenEquals(MethodTable* mt)
+{
+    LIMITED_METHOD_CONTRACT
+
+    if (mt->HasCheckedOverriddenEquals())
+    {
+        return mt->HasOverriddenEquals();
+    }
+
+    BOOL hasOverriddenEquals = FALSE;
+
+    // Do we have overriden Equals method in current Type?
+    if (HasOverriddenEqualsWorker(mt))
+    {
+        hasOverriddenEquals = TRUE;
+    }
+    else
+    {
+        // Now we need to do a depth-first-search on the instance fields.
+        Stack<ApproxFieldDescIterator> fieldIterStack;
+        fieldIterStack.push(ApproxFieldDescIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS));
+        while (!fieldIterStack.empty())
+        {
+            ApproxFieldDescIterator& iter = fieldIterStack.top();
+            FieldDesc* pField = iter.Next();
+            if (pField == NULL)
+            {
+                fieldIterStack.pop();
+                continue;
+            }
+
+            if (pField->GetFieldType() != ELEMENT_TYPE_VALUETYPE
+                && pField->GetFieldType() != ELEMENT_TYPE_CLASS)
+            {
+                // We cannot find overridden Equals in non-class types.
+                continue;
+            }
+
+            // Try to find overriden Equals method in current field type.
+            MethodTable* fieldMethodTable = pField->LookupFieldTypeHandle().GetMethodTable();
+            if (fieldMethodTable->HasCheckedOverriddenEquals())
+            {
+                // If we have checked this field type previously, it implies this field type
+                // does not have overridden Equals.
+                continue;
+            }
+
+            fieldMethodTable->SetHasCheckedOverriddenEquals();
+            if (HasOverriddenEqualsWorker(fieldMethodTable))
```

Comment:
I think `SetHasCheckedOverriddenEquals()` is thread-safe:
https://github.com/mazong1123/coreclr/blob/294536634f3c5256ca10331c1352242edd37712f/src/vm/methodtable.h#L1263

Are you referring from other perspective?